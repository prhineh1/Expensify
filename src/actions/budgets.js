
export const addBudget = (budget = {}, uid) => ({
  type: 'ADD_BUDGET',
  budget,
  uid
});

export const addBudgetComplete = (budget = {}) => ({
  type: 'ADD_BUDGET_COMPLETE',
  budget,
});