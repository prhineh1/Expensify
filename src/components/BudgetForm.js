import React from 'react';

export default class BudgetForm extends React.Component {
  state = {
    description: this.props.budget ? this.props.budget.description : '',
    amount: this.props.budget ? (this.props.budget.amount / 100).toString() : '',
    error: ''
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }))
  }
  onAmountChange = (e) => {
    const amount = e.target.value;
    if(!amount || amount.match(/^((([1-9])\d*)|0)(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }
  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please enter a description and amount.' }));
    }
    else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100
      });
    }
  }
  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          {this.state.error && <p>{this.state.error}</p>}
          <input
            type='text'
            placeholder='Budget Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
           />
          <input
            type="text"
            placehoder='Budgeted Amount'
            value={this.state.amount}
            onChange={this.onAmountChange}
           />
        </form>
      </div>
    )
  }
}