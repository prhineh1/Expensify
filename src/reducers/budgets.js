export default (state = [], action) => {
  switch(action.type) {
    case 'ADD_BUDGET_COMPLETE':
      return [...state, action.budget];
    default:
      return state;
  }
}