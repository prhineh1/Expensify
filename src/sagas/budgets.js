import { takeEvery } from 'redux-saga'
import { put, call } from 'redux-saga/effects';
import * as Api from '../firebase/firebase';

//WATCHERS

export function* watchAddBudgetSaga() {
  yield takeEvery('ADD_BUDGET', addBudgetAsync);
};

//WORKERS

export function* addBudgetAsync({ budget, uid }) {
  try {
    const response = yield call(Api.createBudget, budget, uid);
    yield put({
      type: 'ADD_BUDGET_COMPLETE',
      budget: { id: response.key, ...budget}
    }); 
  }
  catch(error) {
    return;
  }
};