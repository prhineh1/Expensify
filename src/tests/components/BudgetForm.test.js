import React from 'react';
import { shallow } from 'enzyme';
import BudgetForm from '../../components/BudgetForm';
import budgets from '../fixtures/budgets';

test('should render BudgetForm', () => {
  const wrapper = shallow(<BudgetForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render BudgetForm with a budget', () => {
  const wrapper = shallow(<BudgetForm budget={budgets} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<BudgetForm />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot(); 
});

test('should set description on input change', () => {
  const value = 'budget description';
  const wrapper = shallow(<BudgetForm />);
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('description')).toBe(value);  
});

test('should set amount if valid input', () => {
  const value = '45.32';
  const wrapper = shallow(<BudgetForm />);
  wrapper.find('input').at(1).simulate('change', {
      target: { value }
  });
  expect(wrapper.state('amount')).toBe(value);
});

test('should not set amount if invalid input', () => {
  const value = '00.23';
  const wrapper = shallow(<BudgetForm />);
  wrapper.find('input').at(1).simulate('change', {
      target: { value }
  });
  expect(wrapper.state('amount')).toBe('');
});

test('should handle onSubmit for valid form submission', () => {
  const onSubmit = jest.fn();
  const wrapper = shallow(<BudgetForm budget={budgets} onSubmit={onSubmit} />);
  wrapper.find('form').simulate('submit', {
      preventDefault: () => { }
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmit).toHaveBeenLastCalledWith({
      description: budgets.description,
      amount: budgets.amount
  });
});