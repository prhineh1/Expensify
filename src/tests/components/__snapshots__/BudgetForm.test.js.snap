// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BudgetForm 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Budget Description"
      type="text"
      value=""
    />
    <input
      className="text-input"
      onChange={[Function]}
      placehoder="Budgeted Amount"
      type="text"
      value=""
    />
    <div>
      <button
        className="button"
      >
        Save Budget
      </button>
    </div>
  </form>
</div>
`;

exports[`should render BudgetForm with a budget 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Budget Description"
      type="text"
      value="utilities"
    />
    <input
      className="text-input"
      onChange={[Function]}
      placehoder="Budgeted Amount"
      type="text"
      value="200"
    />
    <div>
      <button
        className="button"
      >
        Save Changes
      </button>
    </div>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <p
      className="form__error"
    >
      Please enter a description and amount.
    </p>
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Budget Description"
      type="text"
      value=""
    />
    <input
      className="text-input"
      onChange={[Function]}
      placehoder="Budgeted Amount"
      type="text"
      value=""
    />
    <div>
      <button
        className="button"
      >
        Save Budget
      </button>
    </div>
  </form>
</div>
`;
